### ISSUE TITLE
[R] multi-class demo confusion

### ISSUE THREAD

**Issue by BruceZhaoR** [Microsoft/LightGBM#1682]
https://github.com/Microsoft/LightGBM/blob/425503d7e254130eaea4a5ff287f324f5a5cc1ea/R-package/demo/multiclass_custom_objective.R#L57-L68

Should be  `preds = matrix(preds, nrow = length(labels), byrow = TRUE) ` ?

@jameslamb @StrikerRUS please cc.

**Comment by BruceZhaoR** [comment]
```r
preds <- predict(model_builtin, test[, 1:4])
sum(preds[1:3]) # sum to 1.
preds_mat <- matrix(preds, nrow=nrow(test), byrow = TRUE)
preds_reshape <- predict(model_builtin, test[, 1:4], reshape = TRUE)
identical(preds_mat , preds_reshape )
```
Not really know the custom-eval-metric mechanism,  I appreciate if you would point out my confusion, thanks!

**Comment by Laurae2** [comment]
> Not really know the custom-eval-metric mechanism, the result of eval = custom_multiclass_metric is the same as metric = "multi_logloss".

@BruceZhaoR It is used to show they are identical.

As for the matrix transformation, the memory layout is the following:

```
c = class
p = predicted row
memory layout: c1_p1 c1_p2 c1_p3 ... c2_p1 c2_p2 c2_p3 ...
```

Therefore, `matrix(preds, nrow = length(labels), byrow = FALSE)` should be correct (not `byrow = TRUE`).

It can be checked with the following and running the training script interactively for the 2 lines after the debugging session:

```r
 # define custom metric 
 custom_multiclass_metric = function(preds, dtrain) { 

     labels = getinfo(dtrain, "label") 
     browser()
     print(matrix(preds, nrow = length(labels)))
     print(matrix(preds, nrow = length(labels), byrow = TRUE))
     preds = matrix(preds, nrow = length(labels)) 
     preds = preds - apply(preds, 1, max) 
     prob = exp(preds) / rowSums(exp(preds)) 
  
     return(list(name = "error", 
                 value = -mean(log(prob[cbind(1:length(labels), labels + 1)])), 
                 higher_better = FALSE)) 
  
 } 
```

**Comment by Laurae2** [comment]
Here are the results of the debugging session for reference. Note that labels are aligned by rows, therefore there are 3 groups (1-10, 11-20, and 21-30).

```r
Browse[2]> n
           [,1]      [,2]      [,3]
 [1,]  7.086275 -5.832216 -6.436212
 [2,]  7.086275 -5.665189 -6.436212
 [3,]  7.086275 -5.832231 -6.436212
 [4,]  5.467525 -4.100576 -6.436212
 [5,] -3.141083  4.477595 -6.436216
 [6,]  7.086275 -6.127760 -6.436212
 [7,]  7.086275 -5.832240 -6.436212
 [8,]  7.086275 -5.832231 -6.436212
 [9,]  7.086275 -5.832216 -6.436212
[10,]  7.086275 -5.832231 -6.436212
[11,] -6.070589  6.695462 -6.436213
[12,] -6.190858  6.232891 -6.436216
[13,] -6.322860  6.695462 -6.436213
[14,] -4.891910  6.376302 -6.436213
[15,] -6.322860  6.695462 -6.436213
[16,] -6.190858  6.232891 -6.436216
[17,] -6.322860  6.232891 -6.436213
[18,] -6.322857  6.232891 -6.436213
[19,] -4.891910  6.376302 -6.436213
[20,] -6.322860  6.232891 -6.436213
[21,] -6.190852 -5.104438  6.900235
[22,] -6.190854 -4.667841  6.312166
[23,] -6.322869 -4.532796  6.941599
[24,] -6.190852 -5.104438  6.900235
[25,] -6.190852 -5.104438  6.900235
[26,] -6.190854 -4.667841  6.312166
[27,] -6.322857 -4.212784  6.376878
[28,] -6.190854 -5.266184  6.312166
[29,] -6.190852 -5.407253  7.222196
[30,] -6.190867 -4.995368  6.676697
debug at #6: print(matrix(preds, nrow = length(labels), byrow = TRUE))
Browse[2]> n
           [,1]      [,2]      [,3]
 [1,]  7.086275  7.086275  7.086275
 [2,]  5.467525 -3.141083  7.086275
 [3,]  7.086275  7.086275  7.086275
 [4,]  7.086275 -6.070589 -6.190858
 [5,] -6.322860 -4.891910 -6.322860
 [6,] -6.190858 -6.322860 -6.322857
 [7,] -4.891910 -6.322860 -6.190852
 [8,] -6.190854 -6.322869 -6.190852
 [9,] -6.190852 -6.190854 -6.322857
[10,] -6.190854 -6.190852 -6.190867
[11,] -5.832216 -5.665189 -5.832231
[12,] -4.100576  4.477595 -6.127760
[13,] -5.832240 -5.832231 -5.832216
[14,] -5.832231  6.695462  6.232891
[15,]  6.695462  6.376302  6.695462
[16,]  6.232891  6.232891  6.232891
[17,]  6.376302  6.232891 -5.104438
[18,] -4.667841 -4.532796 -5.104438
[19,] -5.104438 -4.667841 -4.212784
[20,] -5.266184 -5.407253 -4.995368
[21,] -6.436212 -6.436212 -6.436212
[22,] -6.436212 -6.436216 -6.436212
[23,] -6.436212 -6.436212 -6.436212
[24,] -6.436212 -6.436213 -6.436216
[25,] -6.436213 -6.436213 -6.436213
[26,] -6.436216 -6.436213 -6.436213
[27,] -6.436213 -6.436213  6.900235
[28,]  6.312166  6.941599  6.900235
[29,]  6.900235  6.312166  6.376878
[30,]  6.312166  7.222196  6.676697
debug at #7: preds = matrix(preds, nrow = length(labels))
```

**Comment by BruceZhaoR** [comment]
Thank your very much !!
